File: tests/python/test_rvi.py
"""
Python binding tests for RVI indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    # If not in virtual environment, try to import from installed location
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestRvi:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_rvi_partial_params(self, test_data):
        """Test RVI with partial parameters - mirrors check_rvi_partial_params"""
        close = test_data['close']
        
        # Test with partial params (period=10, others default)
        result = ta_indicators.rvi(close, period=10, ma_len=14, matype=1, devtype=0)
        assert len(result) == len(close)
    
    def test_rvi_accuracy(self, test_data):
        """Test RVI matches expected values from Rust tests - mirrors check_rvi_example_values"""
        close = test_data['close']
        
        result = ta_indicators.rvi(close, period=10, ma_len=14, matype=1, devtype=0)
        
        assert len(result) == len(close)
        
        # Check last 5 values match expected from Rust tests
        expected_last_five = [
            67.48579363423423,
            62.03322230763894,
            56.71819195768154,
            60.487299747927636,
            55.022521428674175
        ]
        assert_close(
            result[-5:], 
            expected_last_five,
            rtol=1e-1,  # Using larger tolerance as per Rust test
            msg="RVI last 5 values mismatch"
        )
        
        # Check warmup period has NaN values
        warmup = 10 - 1 + 14 - 1  # period - 1 + ma_len - 1
        for i in range(warmup):
            assert np.isnan(result[i]), f"Expected NaN at index {i} during warmup"
    
    def test_rvi_default_params(self, test_data):
        """Test RVI with default parameters - mirrors check_rvi_default_params"""
        close = test_data['close']
        
        # Default params: period=10, ma_len=14, matype=1, devtype=0
        result = ta_indicators.rvi(close, 10, 14, 1, 0)
        assert len(result) == len(close)
    
    def test_rvi_zero_period(self):
        """Test RVI fails with zero period - mirrors check_rvi_error_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0, 40.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.rvi(input_data, period=0, ma_len=14, matype=1, devtype=0)
    
    def test_rvi_zero_ma_len(self):
        """Test RVI fails with zero ma_len - mirrors check_rvi_error_zero_ma_len"""
        input_data = np.array([10.0, 20.0, 30.0, 40.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.rvi(input_data, period=10, ma_len=0, matype=1, devtype=0)
    
    def test_rvi_period_exceeds_data_length(self):
        """Test RVI fails when period exceeds data length - mirrors check_rvi_error_period_exceeds_data_length"""
        data_small = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.rvi(data_small, period=10, ma_len=14, matype=1, devtype=0)
    
    def test_rvi_all_nan_input(self):
        """Test RVI with all NaN values - mirrors check_rvi_all_nan_input"""
        all_nan = np.full(3, np.nan)
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.rvi(all_nan, period=10, ma_len=14, matype=1, devtype=0)
    
    def test_rvi_not_enough_valid_data(self):
        """Test RVI with not enough valid data - mirrors check_rvi_not_enough_valid_data"""
        data = np.array([np.nan, 1.0, 2.0, 3.0])
        
        with pytest.raises(ValueError, match="Not enough valid data"):
            ta_indicators.rvi(data, period=3, ma_len=5, matype=1, devtype=0)
    
    def test_rvi_empty_input(self):
        """Test RVI fails with empty input"""
        empty = np.array([])
        
        with pytest.raises(ValueError):
            ta_indicators.rvi(empty, period=10, ma_len=14, matype=1, devtype=0)
    
    def test_rvi_batch_single_parameter_set(self, test_data):
        """Test batch with single parameter combination"""
        close = test_data['close']
        
        # Single parameter combination
        batch_result = ta_indicators.rvi_batch(
            close,
            period_range=(10, 10, 0),
            ma_len_range=(14, 14, 0),
            matype_range=(1, 1, 0),
            devtype_range=(0, 0, 0)
        )
        
        # Should match single calculation
        single_result = ta_indicators.rvi(close, 10, 14, 1, 0)
        
        assert batch_result['values'].shape == (1, len(close))
        assert_close(
            batch_result['values'][0], 
            single_result, 
            rtol=1e-10,
            msg="Batch vs single RVI mismatch"
        )
    
    def test_rvi_batch_multiple_periods(self, test_data):
        """Test batch with multiple period values"""
        close = test_data['close'][:100]  # Use smaller dataset for speed
        
        # Multiple periods: 10, 12, 14
        batch_result = ta_indicators.rvi_batch(
            close,
            period_range=(10, 14, 2),
            ma_len_range=(14, 14, 0),
            matype_range=(1, 1, 0), 
            devtype_range=(0, 0, 0)
        )
        
        # Should have 3 rows * 100 cols
        assert batch_result['values'].shape == (3, 100)
        assert len(batch_result['periods']) == 3
        assert batch_result['periods'][0] == 10
        assert batch_result['periods'][1] == 12
        assert batch_result['periods'][2] == 14
        
        # Verify each row matches individual calculation
        periods = [10, 12, 14]
        for i, period in enumerate(periods):
            single_result = ta_indicators.rvi(close, period, 14, 1, 0)
            assert_close(
                batch_result['values'][i], 
                single_result, 
                rtol=1e-10,
                msg=f"Period {period} batch mismatch"
            )
    
    def test_rvi_batch_full_parameter_sweep(self, test_data):
        """Test full parameter sweep"""
        close = test_data['close'][:50]
        
        batch_result = ta_indicators.rvi_batch(
            close,
            period_range=(10, 14, 2),  # 3 periods
            ma_len_range=(10, 12, 2),  # 2 ma_lens
            matype_range=(0, 1, 1),    # 2 matypes
            devtype_range=(0, 1, 1)     # 2 devtypes
        )
        
        # Should have 3 * 2 * 2 * 2 = 24 combinations
        assert batch_result['values'].shape == (24, 50)
        assert len(batch_result['periods']) == 24
        assert len(batch_result['ma_lens']) == 24
        assert len(batch_result['matypes']) == 24
        assert len(batch_result['devtypes']) == 24
        
        # Verify structure
        for row in range(batch_result['values'].shape[0]):
            row_data = batch_result['values'][row]
            
            # Check warmup has NaN
            # Warmup depends on period and ma_len
            period = batch_result['periods'][row]
            ma_len = batch_result['ma_lens'][row]
            warmup = period - 1 + ma_len - 1
            for i in range(min(warmup, len(row_data))):
                if not np.isnan(row_data[i]):
                    # Only check if we expected NaN
                    pass  # Some combinations might have shorter warmup
            
            # After warmup should have values
            for i in range(warmup, len(row_data)):
                assert not np.isnan(row_data[i]), f"Unexpected NaN at index {i} for combination {row}"
    
    def test_rvi_streaming(self):
        """Test RVI streaming functionality"""
        # Create stream with default params
        stream = ta_indicators.RviStream(period=10, ma_len=14, matype=1, devtype=0)
        
        # Update should return None (not fully supported)
        result = stream.update(100.0)
        assert result is None
    
    def test_rvi_kernel_auto(self, test_data):
        """Test RVI with auto kernel selection"""
        close = test_data['close']
        
        result = ta_indicators.rvi(close, 10, 14, 1, 0, kernel="auto")
        assert len(result) == len(close)
    
    def test_rvi_kernel_scalar(self, test_data):
        """Test RVI with scalar kernel"""
        close = test_data['close']
        
        result = ta_indicators.rvi(close, 10, 14, 1, 0, kernel="scalar")
        assert len(result) == len(close)
    
    def test_rvi_invalid_kernel(self, test_data):
        """Test RVI with invalid kernel name"""
        close = test_data['close']
        
        with pytest.raises(ValueError, match="kernel"):
            ta_indicators.rvi(close, 10, 14, 1, 0, kernel="invalid_kernel")
    
    def test_rvi_different_devtypes(self, test_data):
        """Test RVI with different deviation types"""
        close = test_data['close'][:100]  # Smaller dataset
        
        # Test all three devtypes
        devtypes = [0, 1, 2]  # StdDev, MeanAbsDev, MedianAbsDev
        results = []
        
        for devtype in devtypes:
            result = ta_indicators.rvi(close, 10, 14, 1, devtype)
            results.append(result)
        
        # Results should be different for different devtypes
        for i in range(len(devtypes)):
            for j in range(i + 1, len(devtypes)):
                # Check that results are not identical
                diff = np.abs(results[i] - results[j])
                # After warmup, there should be some differences
                assert np.nanmax(diff) > 1e-10, f"Devtype {devtypes[i]} and {devtypes[j]} gave identical results"
    
    def test_rvi_different_matypes(self, test_data):
        """Test RVI with different MA types"""
        close = test_data['close'][:100]  # Smaller dataset
        
        # Test both matypes
        matypes = [0, 1]  # SMA, EMA
        results = []
        
        for matype in matypes:
            result = ta_indicators.rvi(close, 10, 14, matype, 0)
            results.append(result)
        
        # Results should be different for different matypes
        diff = np.abs(results[0] - results[1])
        # After warmup, there should be some differences
        assert np.nanmax(diff) > 1e-10, "SMA and EMA gave identical results"


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/wasm/test_rvi.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {

    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);

    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }

    testData = loadTestData();
});

test('RVI partial params', () => {

    const close = new Float64Array(testData.close);


    const result = wasm.rvi_js(close, 10, 14, 1, 0);
    assert.strictEqual(result.length, close.length);
});

test('RVI accuracy', async () => {

    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.rvi;

    const result = wasm.rvi_js(
        close,
        expected.defaultParams.period,
        expected.defaultParams.ma_len,
        expected.defaultParams.matype,
        expected.defaultParams.devtype
    );

    assert.strictEqual(result.length, close.length);


    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        1e-1,
        "RVI last 5 values mismatch"
    );


    const warmup = expected.defaultParams.period - 1 + expected.defaultParams.ma_len - 1;
    for (let i = 0; i < warmup; i++) {
        assert(isNaN(result[i]), `Expected NaN at index ${i} during warmup`);
    }


    await compareWithRust('rvi', result, 'close', expected.defaultParams);
});

test('RVI default params', () => {

    const close = new Float64Array(testData.close);


    const result = wasm.rvi_js(close, 10, 14, 1, 0);
    assert.strictEqual(result.length, close.length);
});

test('RVI zero period', () => {

    const inputData = new Float64Array([10.0, 20.0, 30.0, 40.0]);

    assert.throws(() => {
        wasm.rvi_js(inputData, 0, 14, 1, 0);
    }, /Invalid period/);
});

test('RVI zero ma_len', () => {

    const inputData = new Float64Array([10.0, 20.0, 30.0, 40.0]);

    assert.throws(() => {
        wasm.rvi_js(inputData, 10, 0, 1, 0);
    }, /Invalid period/);
});

test('RVI period exceeds data length', () => {

    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.rvi_js(dataSmall, 10, 14, 1, 0);
    }, /Invalid period/);
});

test('RVI all NaN input', () => {

    const allNaN = new Float64Array(3).fill(NaN);

    assert.throws(() => {
        wasm.rvi_js(allNaN, 10, 14, 1, 0);
    }, /All values are NaN/);
});

test('RVI not enough valid data', () => {


    const data = new Float64Array([NaN, NaN, NaN, 1.0, 2.0, 3.0, 4.0, 5.0]);

    assert.throws(() => {
        wasm.rvi_js(data, 3, 5, 1, 0);
    }, /Invalid period|Not enough valid data/);
});

test('RVI empty input', () => {

    const empty = new Float64Array([]);

    assert.throws(() => {
        wasm.rvi_js(empty, 10, 14, 1, 0);
    }, /Empty data/);
});

test('RVI batch single parameter set', () => {

    const close = new Float64Array(testData.close);


    const config = {
        period_range: [10, 10, 0],
        ma_len_range: [14, 14, 0],
        matype_range: [1, 1, 0],
        devtype_range: [0, 0, 0]
    };

    const batchResult = wasm.rvi_batch(close, config);
    const singleResult = wasm.rvi_js(close, 10, 14, 1, 0);

    assert.strictEqual(batchResult.rows, 1);
    assert.strictEqual(batchResult.cols, close.length);
    assert.strictEqual(batchResult.values.length, close.length);


    assertArrayClose(
        batchResult.values,
        singleResult,
        1e-10,
        "Batch vs single RVI mismatch"
    );
});

test('RVI batch multiple periods', () => {

    const close = new Float64Array(testData.close.slice(0, 100));


    const config = {
        period_range: [10, 14, 2],
        ma_len_range: [14, 14, 0],
        matype_range: [1, 1, 0],
        devtype_range: [0, 0, 0]
    };

    const batchResult = wasm.rvi_batch(close, config);


    assert.strictEqual(batchResult.rows, 3);
    assert.strictEqual(batchResult.cols, 100);
    assert.strictEqual(batchResult.values.length, 3 * 100);
    assert.strictEqual(batchResult.periods.length, 3);
    assert.strictEqual(batchResult.periods[0], 10);
    assert.strictEqual(batchResult.periods[1], 12);
    assert.strictEqual(batchResult.periods[2], 14);


    const periods = [10, 12, 14];
    for (let i = 0; i < periods.length; i++) {
        const singleResult = wasm.rvi_js(close, periods[i], 14, 1, 0);
        const rowData = batchResult.values.slice(i * 100, (i + 1) * 100);
        assertArrayClose(
            rowData,
            singleResult,
            1e-10,
            `Period ${periods[i]} batch mismatch`
        );
    }
});

test('RVI batch full parameter sweep', () => {

    const close = new Float64Array(testData.close.slice(0, 50));

    const config = {
        period_range: [10, 14, 2],
        ma_len_range: [10, 12, 2],
        matype_range: [0, 1, 1],
        devtype_range: [0, 1, 1]
    };

    const batchResult = wasm.rvi_batch(close, config);


    assert.strictEqual(batchResult.rows, 24);
    assert.strictEqual(batchResult.cols, 50);
    assert.strictEqual(batchResult.values.length, 24 * 50);
    assert.strictEqual(batchResult.periods.length, 24);
    assert.strictEqual(batchResult.ma_lens.length, 24);
    assert.strictEqual(batchResult.matypes.length, 24);
    assert.strictEqual(batchResult.devtypes.length, 24);


    for (let row = 0; row < batchResult.rows; row++) {
        const rowData = batchResult.values.slice(row * 50, (row + 1) * 50);


        const period = batchResult.periods[row];
        const ma_len = batchResult.ma_lens[row];
        const warmup = period - 1 + ma_len - 1;

        for (let i = 0; i < Math.min(warmup, rowData.length); i++) {
            if (!isNaN(rowData[i])) {


            }
        }


        for (let i = warmup; i < rowData.length; i++) {
            assert(!isNaN(rowData[i]), `Unexpected NaN at index ${i} for combination ${row}`);
        }
    }
});


test('RVI zero-copy API', () => {
    const close = new Float64Array(testData.close);
    const len = close.length;


    const inPtr = wasm.rvi_alloc(len);
    const outPtr = wasm.rvi_alloc(len);

    try {

        const expected = wasm.rvi_js(close, 10, 14, 1, 0);


        const inputView = new Float64Array(wasm.__wasm.memory.buffer, inPtr, len);


        inputView.set(close);


        wasm.rvi_into(
            inPtr,
            outPtr,
            len,
            10, 14, 1, 0
        );


        const output = new Float64Array(wasm.__wasm.memory.buffer, outPtr, len);


        const outputArray = Array.from(output);


        assertArrayClose(
            outputArray,
            expected,
            1e-10,
            "Zero-copy RVI mismatch"
        );

    } finally {

        wasm.rvi_free(inPtr, len);
        wasm.rvi_free(outPtr, len);
    }
});

test('RVI zero-copy in-place', () => {

    const data = new Float64Array(testData.close.slice(0, 100));
    const len = data.length;


    const inPtr = wasm.rvi_alloc(len);
    const outPtr = wasm.rvi_alloc(len);

    try {

        new Float64Array(wasm.__wasm.memory.buffer, inPtr, len).set(data);


        wasm.rvi_into(inPtr, outPtr, len, 10, 14, 1, 0);
        const expected = Array.from(new Float64Array(wasm.__wasm.memory.buffer, outPtr, len));


        wasm.rvi_into(inPtr, inPtr, len, 10, 14, 1, 0);
        const inPlace = Array.from(new Float64Array(wasm.__wasm.memory.buffer, inPtr, len));


        assertArrayClose(inPlace, expected, 1e-10, "In-place RVI mismatch");
    } finally {

        wasm.rvi_free(inPtr, len);
        wasm.rvi_free(outPtr, len);
    }
});

test('RVI streaming (not fully supported)', () => {


    const value = 100.0;




    assert.doesNotThrow(() => {


        assert(typeof wasm.rvi_js === 'function');
    });
});

test('RVI different devtypes', () => {

    const data = new Float64Array(testData.close.slice(0, 100));
    const len = data.length;
    const devtypes = [0, 1, 2];

    const inPtr = wasm.rvi_alloc(len);

    const outPtr = wasm.rvi_alloc(len);

    try {
        new Float64Array(wasm.__wasm.memory.buffer, inPtr, len).set(data);

        const results = devtypes.map((dev) => {
            wasm.rvi_into(inPtr, outPtr, len, 10, 14, 1, dev);
            return Array.from(new Float64Array(wasm.__wasm.memory.buffer, outPtr, len));
        });


        for (let i = 0; i < devtypes.length; i++) {
            for (let j = i + 1; j < devtypes.length; j++) {
                let maxDiff = 0;
                for (let k = 0; k < len; k++) {
                    const a = results[i][k];
                    const b = results[j][k];
                    if (!isNaN(a) && !isNaN(b)) {
                        maxDiff = Math.max(maxDiff, Math.abs(a - b));
                    }
                }
                assert(maxDiff > 1e-10, `Devtype ${devtypes[i]} and ${devtypes[j]} gave identical results`);
            }
        }
    } finally {
        wasm.rvi_free(inPtr, len);
        wasm.rvi_free(outPtr, len);
    }
});

test('RVI different matypes', () => {

    const data = new Float64Array(testData.close.slice(0, 100));
    const len = data.length;
    const matypes = [0, 1];

    const inPtr = wasm.rvi_alloc(len);
    const outPtr = wasm.rvi_alloc(len);

    try {
        new Float64Array(wasm.__wasm.memory.buffer, inPtr, len).set(data);

        const results = matypes.map((ma) => {
            wasm.rvi_into(inPtr, outPtr, len, 10, 14, ma, 0);
            return Array.from(new Float64Array(wasm.__wasm.memory.buffer, outPtr, len));
        });

        let maxDiff = 0;
        for (let i = 0; i < len; i++) {
            const a = results[0][i];
            const b = results[1][i];
            if (!isNaN(a) && !isNaN(b)) {
                maxDiff = Math.max(maxDiff, Math.abs(a - b));
            }
        }
        assert(maxDiff > 1e-10, "SMA and EMA gave identical results");
    } finally {
        wasm.rvi_free(inPtr, len);
        wasm.rvi_free(outPtr, len);
    }
});
